### PR TITLE
[Breaking] API changes on getting latency & PRFT info

### PR DESCRIPTION
**Pull request by echoy-harmonicinc**
## Description

This PR added **getProducerReferenceTimeInfo()** and revamped **getPresentationLatency()**

The new API **getProducerReferenceTimeInfo()** will return array of PRFT info, including:
- Source (inband: MPD or MP4 PRFT box)
- Type (encoder or captured)
- Timings (WCA, PTA)

**getPresentationLatency()** will now:
- Analyse data from **getProducerReferenceTimeInfo()**
- Calculate presentation latency
- Priority when there are multiple PRFT metrics available:
  1. MP4 PRFT box & encoder
  2. MP4 PRFT box & captured
  3. MPD & encoder
  4. MPD & captured

## Screenshots (optional)

Screenshot 2021-11-19 at 10 23 07

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to
  not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

File: lib/media/streaming_engine.js [hunk 1]
```diff
@@ -1743,36 +1744,70 @@ shaka.media.StreamingEngine = class {
...
-    return rv;
+          wca=${wca},
+          wca from PRFT=${wcaPrft},
+          pt=${pt},
+          pta=${pta},
+          pta from PRFT=${ptaPrft},
+          type=${type}`);
+
+      mpdPrtInfo = {
+        type: 'mpd',
```
Comment (acheung-harmonicinc) [inline lib/media/streaming_engine.js]: "type" in PRFT context (https://dashif.org/docs/CR-Low-Latency-Live-r8.pdf) is used to denote encoder vs captured, suggest use "inband": true for prft box source, false for mpd source.

File: lib/media/streaming_engine.js [hunk 2]
```diff
@@ -1743,36 +1744,70 @@ shaka.media.StreamingEngine = class {
...
+        pta: [
+          type === 'captured' ? pta : undefined,
+          type === 'encoder' ? pta : undefined,
+        ],
+        wc,
+      };
+
+      mp4PrftInfo = {
+        type: 'prft',
+        wca: [
```
Comment (acheung-harmonicinc) [inline lib/media/streaming_engine.js]: Sugget simplifying the output as follows, then we don't have to assume the ordering in e.g. pta array:
```
[
    {
        type: encoder,
        wca: ...,
        pta: ...,
        inband: true
    },
    {
        type: captured
        wca: ...
        pta: ...
        inband: false
    }
]
```

Also, when inband prft from box is available (with inband attribute in MPD set as true), I doubt if we still need the mpd version. We may just return the prft box one

Comment (echoy-harmonicinc) [reply lib/media/streaming_engine.js]: so there will be maximum 4 combinations correct?

File: lib/media/streaming_engine.js [hunk 3]
```diff
@@ -1743,36 +1744,47 @@ shaka.media.StreamingEngine = class {
...
+          pt=${pt},
+          pta=${pta},
+          pta from PRFT=${ptaPrft},
+          type=${type}`);
+
+      prtData.push({
+        type,
+        wca: hasPrft ? wcaPrft : wca,
+        pta: hasPrft ? ptaPrft: pta,
+        wc,
```
Comment (acheung-harmonicinc) [inline lib/media/streaming_engine.js]: Is the inclusion of wc required?

File: lib/player.js [hunk 4]
```diff
@@ -2758,7 +2742,7 @@ shaka.Player = class extends shaka.util.FakeEventTarget {
       getPlayRate: () => this.playRateController_ ?
         this.playRateController_.getRealRate() : 0,
       getPresentationTime: () => this.playhead_.getTime(),
-      getPresentationLatency: () => this.getPresentationLatency(),
+      getEncoderLatency: () => this.getEncoderLatency(),
```
Comment (acheung-harmonicinc) [inline lib/player.js]: just keep getPresentationLatency, EncoderLatency means the latency delta caused by encoder

Comment (echoy-harmonicinc) [reply lib/player.js]: Yes I'm currently removing it

File: lib/media/streaming_engine.js [hunk 5]
```diff
@@ -1743,36 +1744,104 @@ shaka.media.StreamingEngine = class {
...
       const type = producerReferenceTime.type;
-      const latency = Math.round((wc - wca) - (pt - pta));
-      shaka.log.debug(`computing presentation latency for
+      const wcaPrft = producerReferenceTime.wca_prft;
+      const ptaPrft = producerReferenceTime.pta_prft;
+      const hasPrft = wcaPrft >= 0 && ptaPrft >= 0;
+
+      shaka.log.debug(`Getting ProducerReferenceTimeInfo for
           representation ${repId}, 
           wc=${wc},
```
Comment (acheung-harmonicinc) [inline lib/media/streaming_engine.js]: wc can be removed

File: lib/media/streaming_engine.js [hunk 6]
```diff
@@ -1743,36 +1744,104 @@ shaka.media.StreamingEngine = class {
...
+    if (!prftInfo) {
+      return null;
+    }
+
+    // MP4 PRFT validation
+    // If failed, fallback to MPD
+    const hasPrft = prftInfo.some((prft) => prft.inband === true);
+    const hasEncoder = prftInfo.some((prft) => prft.type === 'encoder');
+
+    const latencies = prftInfo
```
Comment (acheung-harmonicinc) [inline lib/media/streaming_engine.js]: can we replace with sort, then select the first element to make the intention more clear?

File: lib/media/streaming_engine.js [hunk 7]
```diff
@@ -1743,36 +1744,104 @@ shaka.media.StreamingEngine = class {
...
+        .filter((prft) => prft.inband === hasPrft && prft.type)
+        .map((prft) => {
+          const pt = this.playerInterface_.getPresentationTime() * 1000;
+          const {type, wca, pta} = prft;
+          const wallClock = new Date(wca + (pt - pta));
+
+          let latency = Math.round((wallClock.getTime() - wca)
+            - (pt - pta)) / 1000;
+
+          if (prft.inband) {
```
Comment (acheung-harmonicinc) [inline lib/media/streaming_engine.js]: Latency computation should be independent of whether it is inband

File: lib/media/streaming_engine.js [hunk 8]
```diff
@@ -1743,36 +1744,104 @@ shaka.media.StreamingEngine = class {
...
+            latency = Math.round(Date.now() - wallClock.getTime()) / 1000;
```
Comment (acheung-harmonicinc) [inline lib/media/streaming_engine.js]: We should use a common snapshot of Date.now() for fairness

File: lib/media/streaming_engine.js [hunk 9]
```diff
@@ -1743,36 +1744,104 @@ shaka.media.StreamingEngine = class {
...
+   * If encoder latency is not available, fallback to captured latency
+   *
+   * Used in :
+   * - Web Player SDK - Debug overlay
+   * - SimpleLLAbrManager::increaseVideoBitrate
+   * - LiveCatchUpController::getLatency
+   *
+   * @return {Object|null}
+   */
+  getPresentationLatency() {
```
Comment (acheung-harmonicinc) [inline lib/media/streaming_engine.js]: Consider renaming it as getPresentationLatencyInfo if the return type is not a simple numeric latency value

**Review by acheung-harmonicinc — APPROVED** [review]

**Pull request merged**